### PR TITLE
Modified rand_cpu_x86.c to support builtin hardware randomizer on HPE…

### LINKED ISSUE #12903: Add hardware randomizer capability on NonStop x86 architecture
The default implementation of randomization functions do not include the hardware randomizer available on HPE NonStop x86. This is needed because the operating system does not support `/dev/random` or `/dev/urandom` and EGD is not allowed by FIPS.

This applies to version 3.0.0. I am volunteering to implement this capability.

Note: This capability is not available for the NonStop Itanium port.

### PR DESCRIPTION
**Pull request by rsbeckerca**
Modified rand_cpu_x86.c to support builtin hardware randomizer on HPE NonStop.

CLA: Permission is granted by the author to the OpenSSL team to use these modifications.
Fixes #12903

Signed-off-by: Randall S. Becker <rsbecker@nexbridge.com>

### CODE REVIEW

**Comment by rsbeckerca** [comment]
I pushed an amended commit with the correct sign-off for rsbecker@nexbridge.com that has a valid CLA on file. Sorry about the slip-up. The default signature on our x86 system was from a different domain.

File: providers/implementations/rands/seeding/rand_cpu_x86.c
Comment (slontis) [inline providers/implementations/rands/seeding/rand_cpu_x86.c:47]: Wouldn't it make sense to have this as a separate function?

Comment (rsbeckerca) [reply providers/implementations/rands/seeding/rand_cpu_x86.c:47]: > 
> 
> Wouldn't it make sense to have this as a separate function?

Probably. I can reissue the pull request with this change. I was also concerned that the location of the change may have been inappropriate. If it is proper here, I can make this happen tomorrow.

Comment (slontis) [reply providers/implementations/rands/seeding/rand_cpu_x86.c:47]: @paulidale or @mspncp ?

**Comment by levitte** [comment]
> I pushed an amended commit with the correct sign-off for rsbecker@nexbridge.com that has a valid CLA on file. Sorry about the slip-up. The default signature on our x86 system was from a different domain.

Our CLA checker bot looks at the commit author for that sort of thing, not appended commit message headers, so we do have to ask you to amend the commit with `--author=rsbecker@nexbridge.com`.  The only appended header the bot looks at is `CLA:`

File: providers/implementations/rands/seeding/rand_cpu_x86.c
Comment (rsbeckerca) [reply providers/implementations/rands/seeding/rand_cpu_x86.c:48]: Refactored randomizer fetch code into separate method and pushed.

File: providers/implementations/rands/seeding/rand_cpu_x86.c
Comment (paulidale) [inline providers/implementations/rands/seeding/rand_cpu_x86.c:80]: Coding style put a blank line between declarations and code.  Here and elsewhere.

File: providers/implementations/rands/seeding/rand_cpu_x86.c
Comment (paulidale) [inline providers/implementations/rands/seeding/rand_cpu_x86.c:44]: ??? Spurious accident or intentional?

File: providers/implementations/rands/seeding/rand_cpu_x86.c
Comment (paulidale) [inline providers/implementations/rands/seeding/rand_cpu_x86.c:23]: Space after `#` I think.

File: providers/implementations/rands/seeding/rand_cpu_x86.c
Comment (paulidale) [inline providers/implementations/rands/seeding/rand_cpu_x86.c:19]: I'd prefer to see an approach where `get_hardware_random_value` always exists and calls the appropriate code on a per platform basis (i.e. `OPENSSL_ia32_rdseed_bytes`, `OPENSSL_ia32_rdrand_bytes` or `_rdrand64`.  This isolates the conditional code to one place and gives us a generic function that can be used elsewhere if required.  It also gives us cross architecture possibilities.

**Review by paulidale — CHANGES_REQUESTED** [review]
Why would an x86 processor support RDRAND but not support the CPU flags indicating it's usability?  This seems pretty odd to me.

File: providers/implementations/rands/seeding/rand_cpu_x86.c
Comment (rsbeckerca) [reply providers/implementations/rands/seeding/rand_cpu_x86.c:44]: Accidental. Easily removed.

File: providers/implementations/rands/seeding/rand_cpu_x86.c
Comment (rsbeckerca) [reply providers/implementations/rands/seeding/rand_cpu_x86.c:19]: That is actually a tiny bit closer to the original pull request, although I had everything in one procedure. I would be happy to further refactor this but I don't have a good basis of what platforms support the one of `OPENSSL_ia32_rdseed_bytes`, `OPENSSL_ia32_rdrand_bytes`. I can fold the discriminant into a commonly named procedure similar to the `get_hardware_random_value` for other platforms (better name?).

**Comment by rsbeckerca** [comment]
@mattcaswell @levitte I think we're having that Travis issue again. Or did I do something bad?

File: providers/implementations/rands/seeding/rand_cpu_x86.c
Comment (paulidale) [reply providers/implementations/rands/seeding/rand_cpu_x86.c:19]: They are generated by `gen_random` in crypto/x86cpuid.pl and c`rypto/x86_64cpuid.pl`.  They should exist for all x86 and x86_64 platforms.

It is weird that your platform isn't detecting them, I though the CPUID feature for x86/x86_64 was universal by the time these were introduced.

Comment (rsbeckerca) [reply providers/implementations/rands/seeding/rand_cpu_x86.c:19]: I am surprised also, but the platform is not a traditional x86_64. It is Big Endian and has a whole bunch of strangeness. I have a related issue on the Itanium variant that I am working on for this also.

**Comment by rsbeckerca** [comment]
> 
> 
> @mattcaswell @levitte I think we're having that Travis issue again. Or did I do something bad?

It does not look like anything I did, from the Travis CI gcc output.

**Comment by slontis** [comment]
It was a merge error - see latest PR.

**Comment by rsbeckerca** [comment]
> 
> 
> It was a merge error - see latest PR.

Thanks.

File: providers/implementations/rands/seeding/rand_cpu_x86.c
Comment (rsbeckerca) [reply providers/implementations/rands/seeding/rand_cpu_x86.c]: Fixed, rebased, squashed, pushed

File: providers/implementations/rands/seeding/rand_cpu_x86.c
Comment (rsbeckerca) [reply providers/implementations/rands/seeding/rand_cpu_x86.c:64]: Corrected - I hope.

**Comment by rsbeckerca** [comment]
> 
> 
> Why would an x86 processor support RDRAND but not support the CPU flags indicating it's usability? This seems pretty odd to me.

I think I have an (speculative) answer on this. Looking purely at the timing of events, the c99 compiler was developed by one group many years ago. The externalization of the RDRAND ops were added by request to a different group much more recently and likely did not include requested changes to the compiler(s), or approval was not given for that. This may be resolved when we get a more recent compiler made available but I have no timeframe on when that will be - very much outside my control.

**Comment by rsbeckerca** [comment]
@paulidale Is this ready to go? Anything else I need to do?

**Comment by rsbeckerca** [comment]
@paulidale Once PR #12933 is merged, I'll have a fixup for this PR to update the define used for selecting TANDEM. I think that's the last change needed.

**Comment by rsbeckerca** [comment]
@paulidale I have another change ready, but it's perplexing where to go. The NonStop Itanium platform has a similar randomize approach to `rand_unix.c`, although that method has everything folded together and it's suitable. The approach is a real time clock in nanoseconds and a system clock in microseconds. There is a suitable method to combine those without sleeping. I don't really think putting the code into `rand_unix.c` makes a lot of sense without a complete rewrite. And putting it in `rand_cpu_x86.c`, which easy, is somewhat misleading. Do you have a suggestion on this? Essentially, the code is:
```
unsigned long get_sys_random(void) {
        unsigned long long nanos = TS_NANOSECS_() % UINT_MAX ;
        unsigned long long micros = JULIANTIMESTAMP() % UINT_MAX;
        unsigned long long swapped_micros = (micros % 65536LL << 16LL)
                + (micros / 65536LL);
        unsigned long long random = (nanos * swapped_micros) % UINT_MAX;
        return (unsigned long) random;
}
```

**Comment by paulidale** [comment]
There are two places for the timer to go: `rand_tsc.c` is the randomness source based on timers.  We currently do not allow this to be used because we can't quantify the amount of randomness produced.

It can also go in `get_timer_bits()` in `rand_unix.c` -- there is already special case code for Sun and HP/UX where a high resolution timer call is available.  This is called by `rand_pool_add_additional_data()` which mixes things up but doesn't count towards the randomness available.

If the timer is all you've got, you're out of luck for a seed source.  Better make sure that kernel random number generator is working.


As for the function that collects values:
``` C
unsigned long get_sys_random(void) {
        unsigned long long nanos = TS_NANOSECS_() % UINT_MAX ;
        unsigned long long micros = JULIANTIMESTAMP() % UINT_MAX;
        unsigned long long swapped_micros = (micros % 65536LL << 16LL)
                + (micros / 65536LL);
        unsigned long long random = (nanos * swapped_micros) % UINT_MAX;
        return (unsigned long) random;
}
```

My opinion is that this is attempting to obfuscate things and that it isn't producing anything even remotely close to cryptographic level randomness.  The variables nanos and micros will be highly correlated (since they are measuring the same thing).  The word swapping does nothing for the randomness.  Likewise the multiplication.  I'd just use `TS_NANOSECS_()`:
``` C
uint64_t get_timer_bits(void) {
    return TS_NANOSECS_();
}
```

Any entropy that is present will be in the low order bits of the TS_NANOSECS_().  So maybe `TS_NANOSECS_() & 15`.  The exact number of bits is difficult to estimate: it requires data collection and a lengthy statistical analysis and even then it could be wrong.

**Comment by slontis** [comment]
>Any entropy that is present will be in the low order bits of the TS_NANOSECS_(). So maybe TS_NANOSECS_() & 15. The exact number of bits is difficult to estimate: it requires data collection and a lengthy statistical analysis and even then it could be wrong.

Previous experience with this has shown that even some of the lower bits need to get masked out because of repeated patterns in the bits.. As Pauli says - Validating and justifying an entropy source can be very time consuming.

**Comment by kroeckx** [comment]
> > Why would an x86 processor support RDRAND but not support the CPU flags indicating it's usability? This seems pretty odd to me.
> 
> I think I have an (speculative) answer on this. Looking purely at the timing of events, the c99 compiler was developed by one group many years ago. The externalization of the RDRAND ops were added by request to a different group much more recently and likely did not include requested changes to the compiler(s), or approval was not given for that. This may be resolved when we get a more recent compiler made available but I have no timeframe on when that will be - very much outside my control.

I think you're at least misunderstanding the question. All x86
since the pentium in 1993, and all x86_64 CPUs support the CPUID
instruction. That instruction allows us to find out that RDRAND is
available or not. We have assembler support in
crypto/x86_64cpuid.pl that should allow is to detect if RDRAND is
available or not. The compiler is not involved.

It can be that the OS is masking certain CPU features, and that
even when the CPU supports it, we can't see it.

**Comment by rsbeckerca** [comment]
> 
> > Why would an x86 processor support RDRAND but not support the CPU flags indicating it's usability? This seems pretty odd to me. I think I have an (speculative) answer on this. Looking purely at the timing of events, the c99 compiler was developed by one group many years ago. The externalization of the RDRAND ops were added by request to a different group much more recently and likely did not include requested changes to the compiler(s), or approval was not given for that. This may be resolved when we get a more recent compiler made available but I have no timeframe on when that will be - very much outside my control.
> I think you're at least misunderstanding the question. All x86 since the pentium in 1993, and all x86_64 CPUs support the CPUID instruction. That instruction allows us to find out that RDRAND is available or not. We have assembler support in crypto/x86_64cpuid.pl that should allow is to detect if RDRAND is available or not. The compiler is not involved. It can be that the OS is masking certain CPU features, and that even when the CPU supports it, we can't see it.

I understand, but the instructions are not being detected. Perhaps it is because there is no assembler on this platform and the ELF formats are slightly different. There are also two variants, one is x86 (big endian), the other is ia64 (big endian). I've been through the lack of detection of features in many products, and may be the OS, so this is not a surprise. I can't control that part of things.

File: providers/implementations/rands/seeding/rand_cpu_x86.c
Comment (kroeckx) [inline providers/implementations/rands/seeding/rand_cpu_x86.c:69]: I can't find any manual about _rdrand64, but does it take an unsigned long long, or an uint64_t?

**Comment by kroeckx** [comment]
Is there a reason you don't use the assembler? Is it using a
different ABI? Not using the assembler will mean that you don't
use various CPU support like AESNI which will make it slower.

I'm pretty sure x86 only supports little endian, while ia64 can
switch between big and little endian. Maybe what you mean is that
like VOS, all data is stored in big endian, and you have the
compiler convert it on load and store?

File: providers/implementations/rands/seeding/rand_cpu_x86.c
Comment (rsbeckerca) [reply providers/implementations/rands/seeding/rand_cpu_x86.c:69]: > 
> 
> I can't find any manual about _rdrand64, but does it take an unsigned long long, or an uint64_t?

The method is a builtin intrinsic function on the HPE NonStop TNS/X platform. It is documented in their operating system API documentation site only, which is not SEO optimized. It takes a `uint64_t *` (which the compiler is not defining for some reason, but that is equivalent to `unsigned long long *` on the platform).

**Comment by rsbeckerca** [comment]
> 
> Is there a reason you don't use the assembler? Is it using a different ABI? Not using the assembler will mean that you don't use various CPU support like AESNI which will make it slower. I'm pretty sure x86 only supports little endian, while ia64 can switch between big and little endian. Maybe what you mean is that like VOS, all data is stored in big endian, and you have the compiler convert it on load and store?

1. There is no available assembler on the platforms (TNS/E-itanium, or TNS/X-x86), not for a lack of trying to get one for years. gcc is also not available either or we would be using it. While objects are ELF format, they are not standard.
1. The x86 big-endian is implemented in custom millicode on the TNS/X platform.

File: providers/implementations/rands/seeding/rand_cpu_x86.c
Comment (kroeckx) [reply providers/implementations/rands/seeding/rand_cpu_x86.c:69]: We use uint64_t in various files, so it really should be available.

**Comment by rsbeckerca** [comment]
> 
> 
> There are two places for the timer to go: `rand_tsc.c` is the randomness source based on timers. We currently do not allow this to be used because we can't quantify the amount of randomness produced.
> 
> It can also go in `get_timer_bits()` in `rand_unix.c` -- there is already special case code for Sun and HP/UX where a high resolution timer call is available. This is called by `rand_pool_add_additional_data()` which mixes things up but doesn't count towards the randomness available.
> 
> If the timer is all you've got, you're out of luck for a seed source. Better make sure that kernel random number generator is working.
> 
> As for the function that collects values:
> 
> ```c
> unsigned long get_sys_random(void) {
>         unsigned long long nanos = TS_NANOSECS_() % UINT_MAX ;
>         unsigned long long micros = JULIANTIMESTAMP() % UINT_MAX;
>         unsigned long long swapped_micros = (micros % 65536LL << 16LL)
>                 + (micros / 65536LL);
>         unsigned long long random = (nanos * swapped_micros) % UINT_MAX;
>         return (unsigned long) random;
> }
> ```
> 
> My opinion is that this is attempting to obfuscate things and that it isn't producing anything even remotely close to cryptographic level randomness. The variables nanos and micros will be highly correlated (since they are measuring the same thing). The word swapping does nothing for the randomness. Likewise the multiplication. I'd just use `TS_NANOSECS_()`:
> 
> ```c
> uint64_t get_timer_bits(void) {
>     return TS_NANOSECS_();
> }
> ```
> 
> Any entropy that is present will be in the low order bits of the TS_NANOSECS_(). So maybe `TS_NANOSECS_() & 15`. The exact number of bits is difficult to estimate: it requires data collection and a lengthy statistical analysis and even then it could be wrong.

The `TS_NANOSECS_()` method returns nanoseconds since the last system cold-load while `JULIANTIMESTAMP()` returns microseconds since about the creation of the world in biblical tradition. The two are independent. I get that this is hard to certify. There is no explicit hardware randomizer, so doing this based on what is in `rand_unix.c` seems closer to that.

In any event, the current pull request only applies to the x86 platform, and I'll keep it that way.

**Comment by kroeckx** [comment]
From what I read online, it's the compiler (and OS) that converts it to big endian.

The 2 timestamps might have a different starting point and resolutions, but they are most likely generated from the same hardware clock.

**Comment by rsbeckerca** [comment]
> 
> 
> From what I read online, it's the compiler (and OS) that converts it to big endian.
> 
> The 2 timestamps might have a different starting point and resolutions, but they are most likely generated from the same hardware clock.

They are not generated by the same hardware clock. CPU coldload times are real-time clocks in each CPU and are independent. The System clock is an aggregate of the DateTime Clock from all CPUs and synchronises/self-corrects/publishes via the message system. They two are actually very unrelated. The DateTime clock is subject to a large amount of actual drift and it takes regular updates from NIST to keep it sane.

**Comment by rsbeckerca** [comment]
Rechecking uint64_t. Not posting anything relating to Itanium in this pull request. That will be a separate one if we go there.

File: providers/implementations/rands/seeding/rand_cpu_x86.c
Comment (rsbeckerca) [reply providers/implementations/rands/seeding/rand_cpu_x86.c:69]: Pushed a new change using uint64_t. Probably operator error last time.

**Comment by levitte** [comment]
I'm not sure I understand why this PR has issues.  Sure, @openssl/committers are often used to Linux or similar and the offer of a multitude of free compilers / assemblers, and here comes code for a platform very few of us have a good idea of, apart from manuals we can find on the net.  For this section of code, I don't think we have any other option than to trust @rsbecker's knowledge...  and it shouldn't matter for the rest of us, as it's guarded well enough.

And sure, some *future* fixes may do a better job of checking CPUID and find out all sorts of stuff, and starting to use the assembler code we have.  It doesn't have to happen now.  If this code is the difference between "working" and "not working" for NonStop machinery, that should be good enough for now.

**Comment by levitte** [comment]
(it took a looooong time before VMS builds started using the assembler code, at all, FYI...)

**Review by levitte — APPROVED** [review]

**Review by beldmit — APPROVED** [review]
LGTM if Travis is happy.

**Comment by paulidale** [comment]
> The `TS_NANOSECS_()` method returns nanoseconds since the last system cold-load while `JULIANTIMESTAMP()` returns microseconds since about the creation of the world in biblical tradition. The two are independent. 

No they are **not** independent.  The two are _extremely_ correlated: each advances one second per second.  In terms of randomness obtainable, a second timer adds effectively nothing.


I don't believe that [Julian Day Numbers](https://en.wikipedia.org/wiki/Julian_day#Julian_Period) are biblical in origin...

**Review by paulidale — APPROVED** [review]

**Comment by rsbeckerca** [comment]
> 
> 
> > The `TS_NANOSECS_()` method returns nanoseconds since the last system cold-load while `JULIANTIMESTAMP()` returns microseconds since about the creation of the world in biblical tradition. The two are independent.
> 
> No they are **not** independent. The two are _extremely_ correlated: each advances one second per second. In terms of randomness obtainable, a second timer adds effectively nothing.
> 
> I don't believe that [Julian Day Numbers](https://en.wikipedia.org/wiki/Julian_day#Julian_Period) are biblical in origin...

Just for info: on this platform, `JULIANTIMESTAMP() == 0` corresponds to 12:00 UTC on January 1, 4713 B.C. The number is computed software clock aggregated across processors. Each of the 16 logical CPUs has their own real time clock that the 6 IPUs use for the `TS_NANOSECS()`.

Before anything is done for the Itanium box, in a separate PR/Issue, I will be subjecting any random generator that uses these to a heavy weight battery test to assess randomness. This is not happening here. I will communicate the results of the tests. If I cannot get even the NIST tests to pass (a subset of what I am testing), that PR won't see the light of day. Ever. Aside from that, the box (TNS/E) goes off support in 2.5 years, so this may be a moot point if I can't get this to work by then. ;)

**Comment by paulidale** [comment]
Use `Test01`, it's the most comprehensive in my experience.

**Comment by rsbeckerca** [comment]
> 
> 
> Use `Test01`, it's the most comprehensive in my experience.

👍 Already downloaded and installed. That was the plan - using the full battery of tests.

**Comment by paulidale** [comment]
> 👍 Already downloaded and installed. That was the plan - using the full battery of tests.

Make sure you do the tests on the raw entropy, not on the cooked version after your function _whitens_ it.  Test01 I likely to detect the whitening but it isn't certain.

**Comment by kroeckx** [comment]
I find the NIST tool very useful: https://github.com/usnistgov/SP800-90B_EntropyAssessment

**Comment by paulidale** [comment]
> I find the NIST tool very useful: https://github.com/usnistgov/SP800-90B_EntropyAssessment

There are some serious problems with the NIST tooling.  The restart tests e.g. fail a significant percentage of the time even with truly random data.  I can't find the reference to the rebuttal of the tests -- it was presented at ICMC two years back.

**Comment by levitte** [comment]
> There is rarely any correlation even at 10 parts per 10000 between the two clocks with the former being in software and the latter being in hardware.

Without disputing how the two clocks actually work, I would still assume that they both (mostly) increment...  I don't have the deep knowledge of randomness that others have, and still, just that assumption makes red flags go wild in my head.
Generally speaking, using clocks as a source of entropy is considered a bad choice.  I suspect people would otherwise abuse NTP for that sort of thing.

**Comment by kroeckx** [comment]
It's the only tool I know that's designed to test the raw input. All others I know look at the output of the RNG.

**Comment by kroeckx** [comment]
You probably mean: https://www.untruth.org/~josh/sp80090b/Care%20and%20Feeding%20of%20SP800-90B%20D12.pdf

**Comment by paulidale** [comment]
Yes, it was Josh Jill who discovered the worst of it.  I submitted a few comments which were mostly addressed in the final revision.  The committee has been refining things since then.

**Comment by kroeckx** [comment]
There is also this very recent presentation: https://www.untruth.org/~josh/sp80090b/90B%20Extravaganza%2020200921.pdf, and  new repository: https://github.com/KeyPair-Consulting/Theseus

**Comment by rsbeckerca** [comment]
For the Itanium generator, not included in this PR, we got the following results from the small battery on TestU01:
```
       Test                          p-value
 ----------------------------------------------
  4  SimpPoker                        eps
  5  CouponCollector                  eps
  6  MaxOft AD                      1 -  4.3e-5
  7  WeightDistrib                    eps
  9  HammingIndep                     eps
 ----------------------------------------------
 All other tests were passed
```
I'd say "not bad" but "not good enough". So we're in the won't happen soon category for the Itanium RNG at this stage. It should not impact this PR which is x86 only with that hardware RNG that does work.

**Comment by rsbeckerca** [comment]
Although, when running the FIPS-140-2 test:
```
============== Summary results of FIPS-140-2 ==============

 Generator:        nse_gsr
 Number of bits:   20000

       Test          s-value        p-value    FIPS Decision
 --------------------------------------------------------
 Monobit               9854           0.98       Pass
 Poker                14.00           0.53       Pass

 0 Runs, length 1:     2519                      Pass
 0 Runs, length 2:     1242                      Pass
 0 Runs, length 3:      648                      Pass
 0 Runs, length 4:      336                      Pass
 0 Runs, length 5:      165                      Pass
 0 Runs, length 6+:     150                      Pass

 1 Runs, length 1:     2609                      Pass
 1 Runs, length 2:     1248                      Pass
 1 Runs, length 3:      614                      Pass
 1 Runs, length 4:      313                      Pass
 1 Runs, length 5:      137                      Pass
 1 Runs, length 6+:     140                      Pass

 Longest run of 0:       14           0.46       Pass
 Longest run of 1:       13           0.50       Pass
 ----------------------------------------------------------
 All values are within the required intervals of FIPS-140-2
```

**Comment by levitte** [comment]
> It should not impact this PR which is x86 only with that hardware RNG that does work.

This PR is approved as is, so that's not even a discussion at this point.

**Comment by levitte** [comment]
It would probably be fruitful to have the Itanium discussion elsewhere, for separation.

**Comment by rsbeckerca** [comment]
> 
> 
> It would probably be fruitful to have the Itanium discussion elsewhere, for separation.

Thanks. You are quite correct. I have included my analysis in #12988.

**Comment by levitte** [comment]
Merged

d3edef83f5fa378237fcece038f9aff9f89f34cd Modified rand_cpu_x86.c to support builtin hardware randomizer on HPE NonStop.

**Pull request closed without merging**
